src/minisaml/response.py

```diff
@@ -62,11 +75,94 @@ def none(cls) -> "TimeDriftLimits":
         )
 
 
+@dataclass(frozen=True)
+class ValidationConfig:
+    certificate: Union[Certificate, Collection[Certificate]]
+    signature_verification_config: VerifyConfig = VerifyConfig.default()
+    allowed_time_drift: TimeDriftLimits = TimeDriftLimits.none()
+
+
+SyncGetConfigForIssuer = Callable[[str], ValidationConfig]
+AsyncGetConfigForIssuer = Callable[[str], Awaitable[ValidationConfig]]
+
+
+@overload
+def validate_multi_tenant_response(
+    *,
+    data: Union[bytes, str],
+    get_config_for_issuer: SyncGetConfigForIssuer,
+    expected_audience: str,
+) -> Response:
+    pass
+
+
+@overload
+def validate_multi_tenant_response(
+    *,
+    data: Union[bytes, str],
+    get_config_for_issuer: AsyncGetConfigForIssuer,
+    expected_audience: str,
+) -> Awaitable[Response]:
+    pass
+
+
+def validate_multi_tenant_response(
+    *,
+    data: Union[bytes, str],
+    get_config_for_issuer: Union[SyncGetConfigForIssuer, AsyncGetConfigForIssuer],
+    expected_audience: str,
+) -> Union[Response, Awaitable[Response]]:
+    xml = base64.b64decode(data)
+    tree = utils.deserialize_xml(xml)
+    assertion = find_or_raise(tree, "./saml:Assertion")
+    issuer = find_or_raise(assertion, "./saml:Issuer").text
+    maybe_config = get_config_for_issuer(issuer)
```

Comment:
If I'm understanding this right, whoever's calling the API has to also provide the function that gets the config for the issuer? And because that function may or may not be async, we have to handle both options.

Reply:
> If I'm understanding this right, whoever's calling the API has to also provide the function that gets the config for the issuer? And because that function may or may not be async, we have to handle both options.

correct. that's where the crazy `@overload` and the funky action with `asyncio.Future`s comes into play.